### PR TITLE
Fix issue with multiple decoration passes in same Scope

### LINKED ISSUE #328: Multiple decoration passes within same Scope does not work
When a dependency has multiple passes of decorations being applied within the same Module, it does not work.

For example:
```
c := dig.Container()
c.Provide(func () A { ... })
c.Provide(func (a A) B {...})
c.Provide(func (b B) C { ... })

c.Decorate(func() A { ... })
c.Decorate(func(a A) B { ... })

c.Invoke(func(c C) {
   // should see decorators of A and B getting executed, but will only see decorator of B run.
})
```

Fx version of this would be:
```
fx.Module("module",
  fx.Provide(
    func() A { ... },
    func(a A) B { ... },
    func(b B) C { ... },
  ),
  fx.Decorate(
    func() A { ... },
    func(a A) B { ... },
  ),
)
```

This seems to be a regression from one of the bug fixes that were made to Dig. Will shortly follow up with a fix.

### PR DESCRIPTION
**Pull request by sywhang**
This fixes issue when there is a decorator that depends on a type
that is decorated by another decorator that lives in the same
Scope.

Previously, there is a "circuit-breaker" for the parameter building
logic specifically for the case when the decorator depends on the same
type as its output type; this basically causes the decorator.Call and
paramSingle.Build to call each other, ending up in an infinite recursion.

To avoid that, it was skipping over the current Scope when looking for
decorators for its parameters if it was already in the context of a decorator.

However, that logic is incorrect as it is possible for the decorator can still
depend on *other* types that is decorated within the same Scope.

This fixes the issue by keeping track of the decorator's state in 3 modes -
ready, on stack, and called. If a decorator is on stack, that means that it is
still in the middle of being executed, so when parameter builder finds that
decorator, it knows to skip over the current Scope to look for any decorators
from parent Scopes. Otherwise, it can safely look for decorators from the same
Scope since we won't end up in an infinite recursion.

This fixes #328.

### CODE REVIEW

File: param.go [hunk 1]
```diff
@@ -227,14 +227,20 @@ func (ps paramSingle) buildWithDecorators(c containerStore, decorating bool) (v
...
-		if decorators = s.getValueDecorators(ps.Name, ps.Type); len(decorators) > 0 {
+		if decorators = s.getValueDecorators(ps.Name, ps.Type); len(decorators) == 0 {
+			continue
+		}
+		searchParentScope := false
+		for _, d := range decorators {
+			if d.State() == decoratorOnStack {
+				decorators = nil // reset this stuff.
+				searchParentScope = true
+			}
```
Comment (abhinav) [inline param.go:239]: we can probably shorten this by using a labeled continue here:

```
storeLoop:
for _, s := range stores {
   // ..
  for _, d := range decorators {
    if cond {
      decorators = nil
      continue storeLoop
  }
  decoratingScope = s
  break
}
```

**Comment by abhinav** [comment]
Also, is there a reason we stick all decorators in that store instead of the ones that are on-stack? What about something like this?

```diff
diff --git a/param.go b/param.go
index 6d37887..1984030 100644
--- a/param.go
+++ b/param.go
@@ -227,20 +227,21 @@ func (ps paramSingle) buildWithDecorators(c containerStore, decorating bool) (v
 		decoratingScope containerStore
 	)
 	stores := c.storesToRoot()
-storeLoop:
 	for _, s := range stores {
-		if decorators = s.getValueDecorators(ps.Name, ps.Type); len(decorators) == 0 {
-			continue
-		}
-		for _, d := range decorators {
+		decorators = nil
+		for _, d := range s.getValueDecorators(ps.Name, ps.Type) {
 			if d.State() == decoratorOnStack {
-				decorators = nil // reset this stuff.
-				continue storeLoop
+				continue
 			}
+			decorators = append(decorators, d)
+		}
+		if len(decorators) == 0 {
+			continue
 		}
 		decoratingScope = s
 		break
 	}
+
 	if len(decorators) == 0 {
 		return _noValue, false, nil
 	}
```

**Comment by abhinav** [comment]
👍 

RE: value groups: looks like this fails so paramGroupedSlice will need a fix after all

```diff
diff --git a/decorate_test.go b/decorate_test.go
index 15f7e74..38d8d75 100644
--- a/decorate_test.go
+++ b/decorate_test.go
@@ -23,6 +23,7 @@ package dig_test
 import (
 	"errors"
 	"fmt"
+	"strings"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -376,6 +377,46 @@ func TestDecorateSuccess(t *testing.T) {
 		}))
 	})
 
+	t.Run("decorate a value group in the same scope", func(t *testing.T) {
+		type params struct {
+			dig.In
+
+			Strings []string `group:"strings"`
+		}
+
+		type childResult struct {
+			dig.Out
+
+			Strings []string `group:"strings"`
+		}
+
+		type A []string
+
+		parent := digtest.New(t)
+		parent.RequireProvide(func() string { return "a" }, dig.Group("strings"))
+		parent.RequireProvide(func() string { return "b" }, dig.Group("strings"))
+		parent.RequireProvide(func() string { return "c" }, dig.Group("strings"))
+
+		parent.RequireProvide(func(p params) A { return A(p.Strings) })
+
+		child := parent.Scope("child")
+
+		child.RequireDecorate(func(p params) childResult {
+			res := childResult{Strings: make([]string, len(p.Strings))}
+			for i, s := range p.Strings {
+				res.Strings[i] = strings.ToUpper(s)
+			}
+			return res
+		})
+		child.RequireDecorate(func(p params) A {
+			return append(A(p.Strings), "D")
+		})
+
+		require.NoError(t, child.Invoke(func(a A) {
+			assert.ElementsMatch(t, A{"A", "B", "C", "D"}, a)
+		}))
+	})
+
 	t.Run("decorate a value group with an empty slice", func(t *testing.T) {
 		type A struct {
 			dig.In
```

File: param.go [hunk 2]
```diff
@@ -227,17 +227,19 @@ func (ps paramSingle) buildWithDecorators(c containerStore, decorating bool) (v
 		decoratingScope containerStore
 	)
 	stores := c.storesToRoot()
-	// If we are already in a decorating stack,
-	// skip the current Scope to avoid infinite
-	// recursion.
-	if decorating {
```
[the remark this answers is not in the document]
Comment (sywhang) [reply param.go:233]: still being used in paramGroupedSlice. Will remove it after I fix this issue in paramGroupedSlice.

**Pull request merged**